### PR TITLE
[consensus, dag] feat: Wire the DAG in the block removal

### LINKED ISSUE #208: [remove_collections] wire in `dag` in `block_removal`
### Story

Parent story: https://github.com/MystenLabs/narwhal/issues/70

### Description

Currently the `block_remover` is performing the clean up of removed collections from:
* the `certificate_store` (primary)
* the `header_store` (primary)
* the `payload_store` (primary)
* the `batches_store` (worker)

To complete the clean up process we should remove the collections/certificates from the virtual DAG as well. As part of this story we should wire in the virtual dag (technical approach is open to the implementation) in order to perform the clean up there as well.

The suggested place to perform the call to clean up the DAG can be found here:

https://github.com/MystenLabs/narwhal/blob/main/primary/src/block_remover.rs#L361-L363

### PR DESCRIPTION
**Pull request by huitseeker**
Fixes #208

### CODE REVIEW

**Review by asonnino — APPROVED** [review]
LGTM

File: primary/src/block_remover.rs [hunk 1]
```diff
@@ -173,6 +188,9 @@ pub struct BlockRemover<PublicKey: VerifyingKey> {
     /// The persistent storage for payload markers from workers.
     payload_store: Store<(BatchDigest, WorkerId), PayloadToken>,
 
+    /// The Dag structure for managing the stored certificates
+    dag: Option<Arc<Dag<PublicKey>>>,
```
Comment (akichidis) [inline primary/src/block_remover.rs:192]: I see that the approach we take with `Store` is to clone them every time we pass around to other structs. Here we are passing an `Arc` and I see the point (I guess we could go with the `clone` approach as well). What would be the best/recommended approach in general in this case? For example, would we want ideally to share the Stores via Arcs as well (or use copy for Dag)?

**Review by akichidis — APPROVED** [review]
Looks great! Thanks a lot @huitseeker 👍

File: consensus/src/dag.rs [hunk 2]
```diff
@@ -219,13 +207,43 @@ impl<PublicKey: VerifyingKey> InnerDag<PublicKey> {
         self.read_causal(*start_digest)
     }
 
-    /// Removes a certificate from the Dag, reclaiming memory in the process.
-    fn remove(&mut self, digest: CertificateDigest) -> Result<(), ValidatorDagError<PublicKey>> {
+    /// Removes certificates from the Dag, reclaiming memory in the process.
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply consensus/src/dag.rs:210]: That's actually not the case : the removals will occur for every known certificate individually. I've added a comment to explain.

**Pull request merged**